server/core/src/rodsAgent.cpp

```diff
@@ -192,437 +224,529 @@ void cleanup()
-                }
+        const auto agent_pid = fork();
+
+        // This socket will not be used by the agent factory or agent, so close it.
+        close(listen_tmp_socket);
+
+        if (agent_pid == 0) {
+            // This is the child process.
+            // Agent logic starts outside of the while-loop.
+            break;
+        }
+        else if (agent_pid > 0) {
+            // This is the parent process.
+            // Clean up and prepare to fork more agents upon request.
+            close(conn_tmp_socket);
+        }
+        else if (agent_pid < 0) {
+            // The call to fork() failed.
+
+            log_agent_factory::error("fork() failed in rodsAgent process factory.");
+
+            // TODO Why do we shut down everything here?
+            // This feels unnecessary. If the agent factory isn't working correctly, the administrator
+            // can choose to restart the server. Shutting the agent factory down in this way just causes
+            // the main server process to fork a new one.
+
+            if (close(client_socket) < 0) {
+                log_agent_factory::error("close(conn_socket) failed with errno = [{}]: {}", errno, strerror(errno));
+            }
 
-                return SYS_FORK_ERROR;
+            if (close(listen_socket) < 0) {
+                log_agent_factory::error("close(listen_socket) failed with errno = [{}]: {}", errno, strerror(errno));
             }
+
+            return SYS_FORK_ERROR;
+        }
+    } // Agent factory main loop.
+
+    //
+    // This is where the agent logic actually begins.
+    //
+
+    log_agent::trace("Agent forked. Initializing ...");
+
+    close(listen_socket);
+
+    // This seems to have fixed the issue with agents not being reaped by the agent factory.
+    // However, if we don't restore the disposition of SIGABRT, every agent will produce a stacktrace on exit.
+    // The current question is, should agents have custom signal handlers?
+    std::signal(SIGABRT, SIG_DFL);
+    //std::signal(SIGABRT, [](int _signum) { log_agent::error("Agent caught SIGABRT [{}]!", _signum); _exit(_signum); }); // Agents are definitely hitting this!
+    std::signal(SIGINT,  SIG_DFL);
+    std::signal(SIGHUP,  SIG_DFL);
+    std::signal(SIGTERM, SIG_DFL);
+    std::signal(SIGCHLD, SIG_DFL);
+    std::signal(SIGUSR1, SIG_DFL);
+    std::signal(SIGPIPE, SIG_DFL);
+
+    int status{};
+
+    try {
+        log::set_server_type("agent");
+
+        // Reload server_config.json into memory for the newly forked agent process.
+        irods::environment_properties::instance().capture();
+
+        // should we check this return code??
+        status = receiveDataFromServer(conn_tmp_socket);
+        if (status < 0) {
+            log_agent::error("receiveDataFromServer failed [error_code=[{}]].", status);
+            //return err.code(); // TODO Should we return here?
```

Comment:
Looks like this hasn't changed, but there is a new TODO.

Reply:
Done.